include/ginkgo/core/preconditioner/batch_ilu.hpp

```diff
@@ -95,41 +76,64 @@ class BatchIlu : public EnableBatchLinOp<BatchIlu<ValueType, IndexType>>,
+         *        matrix (only skip if it is known that it is already sorted).
+         *
+         * The algorithm to generate exact ilu0 requires the
+         * input matrix to be sorted. If it is, this parameter can be set to
+         * `true` to skip the sorting for better performance.
          */
-        batch_trsv_type GKO_FACTORY_PARAMETER_SCALAR(trsv_type,
-                                                     batch_trsv_type::exact);
+        bool GKO_FACTORY_PARAMETER_SCALAR(skip_sorting, false);
 
         /**
-         * Factorization algorithm to use.
+         * @brief Number of sweeps for parilu0
+         *
+         * This parameter is used only in case of parilu0 preconditioner.
          */
-        batch_factorization_type GKO_FACTORY_PARAMETER_SCALAR(
-            factorization_type, batch_factorization_type::exact);
+        int GKO_FACTORY_PARAMETER_SCALAR(parilu_num_sweeps, 10);
 
         /**
-         * How the factors are stored.
+         * @brief batch_ilu_type
          *
-         * Two separate matrices for L and U (split) or one combined matrix
-         * with implied unit diagonal (combined).
          */
-        batch_factors_storage GKO_FACTORY_PARAMETER_SCALAR(
-            storage_type, batch_factors_storage::split);
+        batch_ilu_type GKO_FACTORY_PARAMETER_SCALAR(ilu_type,
+                                                    batch_ilu_type::exact_ilu);
     };
     GKO_ENABLE_BATCH_LIN_OP_FACTORY(BatchIlu, parameters, Factory);
     GKO_ENABLE_BUILD_METHOD(Factory);
 
-    std::unique_ptr<BatchLinOp> transpose() const override;
 
-    std::unique_ptr<BatchLinOp> conj_transpose() const override;
+    std::unique_ptr<BatchLinOp> transpose() const override
+    {
+        return build()
+            .with_skip_sorting(this->parameters_.skip_sorting)
+            .with_ilu_type(this->parameters_.ilu_type)
+            .with_parilu_num_sweeps(this->parameters_.parilu_num_sweeps)
+            .on(this->get_executor())
+            ->generate(share(
+                as<BatchTransposable>(this->system_matrix_)->transpose()));
+    }
+
+    std::unique_ptr<BatchLinOp> conj_transpose() const override
+    {
+        return build()
+            .with_skip_sorting(this->parameters_.skip_sorting)
+            .with_ilu_type(this->parameters_.ilu_type)
+            .with_parilu_num_sweeps(this->parameters_.parilu_num_sweeps)
+            .on(this->get_executor())
+            ->generate(share(
+                as<BatchTransposable>(this->system_matrix_)->conj_transpose()));
+    }
 
-    const matrix::BatchCsr<ValueType, IndexType>* get_const_lower_factor() const
+    const matrix::BatchCsr<ValueType, IndexType>* get_const_factorized_matrix()
+        const
     {
-        return l_factor_.get();
+        return mat_factored_.get();
     }
 
-    const matrix::BatchCsr<ValueType, IndexType>* get_const_upper_factor() const
+    const IndexType* get_const_diag_locations() const
```

Comment:
Is this needed in the public interface ? If not you can move it to `protected`.

Reply:
Yes, this is needed in the public interface, this function is being used in the tests.